File: server/internal/modules/token_gateway/service/video_redis_capacity.go
package service

import (
	"context"
	"crypto/rand"
	_ "embed"
	"encoding/hex"
	"encoding/json"
	"errors"
	"regexp"
	"strconv"
	"time"

	"github.com/redis/go-redis/v9"
	video "molin/server/internal/modules/token_gateway/video"
)

var (
	ErrVideoCapacityUnavailable = errors.New("视频容量协调暂不可用")
	ErrVideoCapacityConflict    = errors.New("视频容量意图或尝试代次冲突")
	ErrVideoCapacityLeaseLost   = errors.New("视频容量持有者已失效")
	ErrVideoCapacityFull        = errors.New("视频容量已满")
)

const videoCapacityStateKey = "molin:{video-g7}:capacity:active-v1"

// 本组件只操作Redis容量快照，不负责用户授权、MySQL提交证明或Provider调用。
type VideoCapacityIdentity struct {
	TaskID, RequestID          string
	UserID, ProjectID          uint64
	APIKeyID                   *uint64
	Model, Provider, Operation string
}

// 主体ID必须是十进制字符串，防止Lua/cjson把uint64舍入到同一主体。
type videoCapacityIdentityJSON struct {
	Task      string `json:"task"`
	Request   string `json:"request"`
	User      string `json:"user"`
	Project   string `json:"project"`
	Key       string `json:"key"`
	Model     string `json:"model"`
	Provider  string `json:"provider"`
	Operation string `json:"operation"`
}

type VideoCapacityAttempt struct {
	identity string
	task     string
	nonce    string
}

// 同一网络尝试保留这个不可变对象；不能在结果未知时另造nonce覆盖旧预留。
func NewVideoCapacityAttempt(identity VideoCapacityIdentity) (*VideoCapacityAttempt, error) {
	return newVideoCapacityAttempt(identity, "")
}

func newVideoCapacityAttempt(identity VideoCapacityIdentity, fixedNonce string) (*VideoCapacityAttempt, error) {
	body, err := canonicalVideoCapacityIdentity(identity)
	if err != nil {
		return nil, err
	}
	nonce := fixedNonce
	if nonce == "" {
		secret := make([]byte, 32)
		if _, err := rand.Read(secret); err != nil {
			return nil, ErrVideoCapacityUnavailable
		}
		nonce = hex.EncodeToString(secret)
	} else if !regexp.MustCompile(`^[0-9a-f]{64}$`).MatchString(nonce) {
		return nil, ErrVideoCapacityConflict
	}
	return &VideoCapacityAttempt{identity: body, task: identity.TaskID, nonce: nonce}, nil
}

func canonicalVideoCapacityIdentity(identity VideoCapacityIdentity) (string, error) {
	if identity.UserID == 0 || identity.ProjectID == 0 || !videoBillingPublicID.MatchString(identity.TaskID) || !videoBillingPublicID.MatchString(identity.RequestID) || !regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._/-]{0,190}$`).MatchString(identity.Model) || identity.Provider != "fake-native-async" || (identity.Operation != "text_to_video" && identity.Operation != "image_to_video") {
		return "", ErrVideoCapacityConflict
	}
	user, project := strconv.FormatUint(identity.UserID, 10), strconv.FormatUint(identity.ProjectID, 10)
	key := "jwt:" + user + ":" + project
	if identity.APIKeyID != nil {
		if *identity.APIKeyID == 0 {
			return "", ErrVideoCapacityConflict
		}
		key = "sk:" + strconv.FormatUint(*identity.APIKeyID, 10)
	}
	body, err := json.Marshal(videoCapacityIdentityJSON{Task: identity.TaskID, Request: identity.RequestID, User: user, Project: project, Key: key, Model: identity.Model, Provider: identity.Provider, Operation: identity.Operation})
	if err != nil {
		return "", ErrVideoCapacityConflict
	}
	return string(body), nil
}

// 值接收者同时保护指针与解引用副本，内部尝试能力不进入普通JSON或格式化日志。
func (VideoCapacityAttempt) MarshalJSON() ([]byte, error) { return []byte(`{"redacted":true}`), nil }
func (VideoCapacityAttempt) String() string               { return "[video capacity attempt]" }
func (VideoCapacityAttempt) GoString() string             { return "[video capacity attempt]" }

type VideoCapacityView struct {
	Phase                 string
	ExpiresAt, ObservedAt time.Time
	Expired               bool
}

type VideoCapacityLimitError struct {
	Scope      string
	RetryAfter time.Duration
}

func (*VideoCapacityLimitError) Error() string { return ErrVideoCapacityFull.Error() }
func (*VideoCapacityLimitError) Unwrap() error { return ErrVideoCapacityFull }

// 构造器不会初始化空Redis，也不会宣布恢复完成。epoch必须来自后续持久化协调器。
type RedisVideoCapacityStore struct {
	client *redis.Client
	epoch  string
	policy string
	limits [10]uint32
}

func NewRedisVideoCapacityStore(client *redis.Client, epoch uint64, policy *video.VideoCapacityPolicy) (*RedisVideoCapacityStore, error) {
	if client == nil || epoch == 0 || client.Options().MaxRetries != 0 || !client.Options().ContextTimeoutEnabled {
		return nil, ErrVideoCapacityUnavailable
	}
	limits, err := policy.Limits()
	if err != nil {
		return nil, err
	}
	hash, err := policy.Fingerprint()
	if err != nil {
		return nil, err
	}
	return &RedisVideoCapacityStore{client: client, epoch: strconv.FormatUint(epoch, 10), policy: hash, limits: [10]uint32{limits.Queued.User, limits.Queued.Project, limits.Queued.APIKey, limits.Queued.Model, limits.Queued.Global, limits.Running.User, limits.Running.Project, limits.Running.APIKey, limits.Running.Model, limits.Running.Provider}}, nil
}

func (s *RedisVideoCapacityStore) ReserveQueued(ctx context.Context, attempt *VideoCapacityAttempt) (*VideoCapacityView, error) {
	return s.execute(ctx, "reserve", attempt)
}

func (s *RedisVideoCapacityStore) Read(ctx context.Context, attempt *VideoCapacityAttempt) (*VideoCapacityView, error) {
	return s.execute(ctx, "read", attempt)
}

// 只预留running并保留queued；后续数据库提交协调器尚未接线，不能据此调用Provider。
func (s *RedisVideoCapacityStore) PrepareRunning(ctx context.Context, attempt *VideoCapacityAttempt) (*VideoCapacityView, error) {
	return s.execute(ctx, "prepare", attempt)
}

func (s *RedisVideoCapacityStore) Renew(ctx context.Context, attempt *VideoCapacityAttempt) (*VideoCapacityView, error) {
	return s.execute(ctx, "renew", attempt)
}

// confirmRunning只完成已预留promoting到running的原子切换；只能由同包持久协调器在MySQL提交后调用。
func (s *RedisVideoCapacityStore) confirmRunning(ctx context.Context, attempt *VideoCapacityAttempt) (*VideoCapacityView, error) {
	return s.execute(ctx, "confirm", attempt)
}

// releaseCapacity只移除完全匹配的容量记录；只能由同包协调器先证明安全终态、Provider结束和财务事实。
func (s *RedisVideoCapacityStore) releaseCapacity(ctx context.Context, attempt *VideoCapacityAttempt) (*VideoCapacityView, error) {
	return s.execute(ctx, "release", attempt)
}

// abortPromotion只用于MySQL明确未提交时把同一promoting退回queued；提交结果未知时禁止调用。
func (s *RedisVideoCapacityStore) abortPromotion(ctx context.Context, attempt *VideoCapacityAttempt) (*VideoCapacityView, error) {
	return s.execute(ctx, "abort", attempt)
}

// Lua仅返回低敏观察；该结果不能代替MySQL提交证明、Worker围栏或Provider授权。
func (s *RedisVideoCapacityStore) execute(ctx context.Context, action string, attempt *VideoCapacityAttempt) (*VideoCapacityView, error) {
	if s == nil || s.client == nil || ctx == nil || attempt == nil || attempt.identity == "" || attempt.task == "" || !regexp.MustCompile(`^[0-9a-f]{64}$`).MatchString(attempt.nonce) {
		return nil, ErrVideoCapacityUnavailable
	}
	bounded, cancel := context.WithTimeout(ctx, time.Second)
	defer cancel()
	limits, err := json.Marshal(s.limits)
	if err != nil {
		return nil, ErrVideoCapacityUnavailable
	}
	// 客户端构造时已禁用自动重试。结果未知只留给原尝试查询，不更换nonce猜测重新预留。
	values, err := s.client.Eval(bounded, videoCapacityLua, []string{videoCapacityStateKey}, action, s.epoch, s.policy, attempt.task, attempt.identity, attempt.nonce, string(limits)).Slice()
	if err != nil || len(values) < 1 {
		return nil, ErrVideoCapacityUnavailable
	}
	code, ok := values[0].(int64)
	if !ok {
		return nil, ErrVideoCapacityUnavailable
	}
	switch code {
	case 0:
		return nil, ErrVideoCapacityUnavailable
	case 2:
		return nil, ErrVideoCapacityConflict
	case 3:
		return nil, ErrVideoCapacityLeaseLost
	case 4:
		if len(values) != 2 {
			return nil, ErrVideoCapacityUnavailable
		}
		scope, ok := values[1].(string)
		if !ok || (scope != "user" && scope != "project" && scope != "api_key" && scope != "model" && scope != "global" && scope != "provider") {
			return nil, ErrVideoCapacityUnavailable
		}
		return nil, &VideoCapacityLimitError{Scope: scope, RetryAfter: time.Second}
	case 1:
		if len(values) != 4 {
			return nil, ErrVideoCapacityUnavailable
		}
		phase, phaseOK := values[1].(string)
		expires, expiresOK := values[2].(int64)
		now, nowOK := values[3].(int64)
		if !phaseOK || !expiresOK || !nowOK || expires <= 0 || now <= 0 || (phase != "queued" && phase != "promoting" && phase != "running" && phase != "released") {
			return nil, ErrVideoCapacityUnavailable
		}
		return &VideoCapacityView{Phase: phase, ExpiresAt: time.UnixMilli(expires).UTC(), ObservedAt: time.UnixMilli(now).UTC(), Expired: expires <= now}, nil
	default:
		return nil, ErrVideoCapacityUnavailable
	}
}

// 脚本与Go源码一同冻结和计算哈希，便于审查所有读取检查均先于唯一SET。
//
//go:embed video_redis_capacity.lua
var videoCapacityLua string

File: server/internal/modules/token_gateway/service/video_redis_capacity.lua
-- 单一有界快照避免多桶分别增减出现部分事实；空库和损坏状态绝不自动初始化。
local function decode(raw)
  if type(raw) ~= 'string' then return nil end
  local ok, value = pcall(cjson.decode, raw)
  if not ok or type(value) ~= 'table' then return nil end
  return value
end
local function fields(value, names)
  if type(value) ~= 'table' then return false end
  local allowed, size = {}, 0
  for _, name in ipairs(names) do allowed[name] = true end
  for name, _ in pairs(value) do
    if not allowed[name] then return false end
    size = size + 1
  end
  return size == #names
end
local function uint_string(value)
  return type(value) == 'string' and string.match(value, '^[1-9]%d*$') ~= nil
    and (#value < 20 or (#value == 20 and value <= '18446744073709551615'))
end
local function public_id(value)
  return type(value) == 'string' and #value <= 128 and string.match(value, '^[A-Za-z0-9][A-Za-z0-9_.%-]*$') ~= nil
end
local function identity(raw)
  if type(raw) ~= 'string' or #raw > 2048 then return nil end
  local v = decode(raw)
  if not fields(v, {'task','request','user','project','key','model','provider','operation'}) then return nil end
  if not public_id(v.task) or not public_id(v.request) or not uint_string(v.user) or not uint_string(v.project) then return nil end
  if type(v.key) ~= 'string' then return nil end
  local sk = string.match(v.key, '^sk:([1-9]%d*)$')
  if not (sk and uint_string(sk)) and v.key ~= 'jwt:' .. v.user .. ':' .. v.project then return nil end
  if type(v.model) ~= 'string' or #v.model > 191 or not string.match(v.model, '^[A-Za-z0-9][A-Za-z0-9._/%-]*$') then return nil end
  if v.provider ~= 'fake-native-async' or (v.operation ~= 'text_to_video' and v.operation ~= 'image_to_video') then return nil end
  return v
end
local action, epoch, policy, task, raw_identity, nonce = ARGV[1], ARGV[2], ARGV[3], ARGV[4], ARGV[5], ARGV[6]
if action ~= 'reserve' and action ~= 'read' and action ~= 'prepare' and action ~= 'renew' and action ~= 'confirm' and action ~= 'release' and action ~= 'abort' then return {0} end
local target = identity(raw_identity)
if not target or target.task ~= task or type(nonce) ~= 'string' or #nonce ~= 64 or not string.match(nonce, '^[0-9a-f]+$') then return {0} end
local limits = decode(ARGV[7])
local ceilings = {2,10,2,100,100,1,2,1,2,2}
if not limits or #limits ~= 10 then return {0} end
for i = 1, 10 do
  if type(limits[i]) ~= 'number' or limits[i] % 1 ~= 0 or limits[i] < 1 or limits[i] > ceilings[i] then return {0} end
end
local kind = redis.call('TYPE', KEYS[1]).ok
if kind ~= 'string' or redis.call('PTTL', KEYS[1]) ~= -1 then return {0} end
if redis.call('STRLEN', KEYS[1]) > 131072 then return {0} end
local raw = redis.call('GET', KEYS[1])
if not raw or #raw > 131072 then return {0} end
local state = decode(raw)
if not fields(state, {'schema','epoch','policy','run_id','status','count','records'}) then return {0} end
if state.schema ~= 1 or state.epoch ~= epoch or state.policy ~= policy or state.status ~= 'ready'
  or type(state.records) ~= 'table' or type(state.count) ~= 'number' or state.count % 1 ~= 0 or state.count < 0 or state.count > 102 then return {0} end
-- 在实际执行脚本的连接内复验实例身份；重启保留下来的旧ready快照不能授权新写入。
local actual_run_id = string.match(redis.call('INFO', 'server'), 'run_id:([0-9a-f]+)')
if not actual_run_id or #actual_run_id ~= 40 or state.run_id ~= actual_run_id then return {0} end
local clock = redis.call('TIME')
local now = tonumber(clock[1]) * 1000 + math.floor(tonumber(clock[2]) / 1000)
local counts, hard_counts = {0,0,0,0,0,0,0,0,0,0}, {}
local count, request_owners = 0, {}
local target_scopes = {target.user,target.project,target.key,target.model,'global',target.user,target.project,target.key,target.model,target.provider}
for id, row in pairs(state.records) do
  count = count + 1
  if count > 102 or not fields(row, {'identity','attempt','phase','expires_ms'}) then return {0} end
  local subject = identity(row.identity)
  if not subject or subject.task ~= id or type(row.attempt) ~= 'string' or #row.attempt ~= 64 or not string.match(row.attempt, '^[0-9a-f]+$') then return {0} end
  if request_owners[subject.request] then return {0} end
  request_owners[subject.request] = id
  if row.phase ~= 'queued' and row.phase ~= 'promoting' and row.phase ~= 'running' then return {0} end
  if type(row.expires_ms) ~= 'number' or row.expires_ms % 1 ~= 0 or row.expires_ms <= 0 or row.expires_ms > now + 30000 then return {0} end
  local scopes = {subject.user,subject.project,subject.key,subject.model,'global',subject.user,subject.project,subject.key,subject.model,subject.provider}
  for i = 1,10 do
    if (i <= 5 and row.phase ~= 'running') or (i > 5 and row.phase ~= 'queued') then
      local bucket = tostring(i) .. ':' .. scopes[i]
      hard_counts[bucket] = (hard_counts[bucket] or 0) + 1
      if hard_counts[bucket] > ceilings[i] then return {0} end
      if scopes[i] == target_scopes[i] then counts[i] = counts[i] + 1 end
    end
  end
end
if count ~= state.count then return {0} end
if request_owners[target.request] and request_owners[target.request] ~= task then return {2} end
local row = state.records[task]
if row then
  if row.identity ~= raw_identity or row.attempt ~= nonce then return {2} end
  -- Read允许观察过期债务；重复申请不续期，过期持有者不得恢复写权限。
  -- Release由上层持久终态证明授权，允许清理过期债务；其余旧持有者不能恢复写权限。
  if action ~= 'read' and action ~= 'release' and row.expires_ms <= now then return {3} end
  if action == 'read' or action == 'reserve' then return {1,row.phase,row.expires_ms,now} end
  if action == 'prepare' then
    if row.phase ~= 'queued' then return {1,row.phase,row.expires_ms,now} end
    local names = {'user','project','api_key','model','provider'}
    for i = 6,10 do if counts[i] >= limits[i] then return {4,names[i-5]} end end
    -- 预留全部running后仍保留queued；没有MySQL提交证明时不减少排队占用。
    row.phase = 'promoting'
  elseif action == 'confirm' then
    if row.phase == 'running' then return {1,row.phase,row.expires_ms,now} end
    if row.phase ~= 'promoting' then return {3} end
    -- MySQL状态已确定提交后才由上层调用；切到running同时移除queued计数并刷新技术租期。
    row.phase = 'running'
    row.expires_ms = now + 30000
  elseif action == 'renew' then
    row.expires_ms = now + 30000
  elseif action == 'abort' then
    if row.phase == 'queued' then return {1,row.phase,row.expires_ms,now} end
    if row.phase ~= 'promoting' then return {3} end
    -- 仅撤销尚未提交的running预留，保留原queued债务和原技术截止，不借机续期。
    row.phase = 'queued'
  elseif action == 'release' then
    state.records[task] = nil
    state.count = count - 1
  end
else
  -- 同一安全终态释放的重放是只读成功；不存在记录不能为其他动作凭空创建许可。
  if action == 'release' then return {1,'released',now,now} end
  if action ~= 'reserve' then return {3} end
  local names = {'user','project','api_key','model','global'}
  for i = 1,5 do if counts[i] >= limits[i] then return {4,names[i]} end end
  if count >= 102 then return {0} end
  row = {identity=raw_identity,attempt=nonce,phase='queued',expires_ms=now+30000}
  state.records[task], state.count = row, count + 1
end
local ok, changed = pcall(cjson.encode, state)
if not ok or #changed > 131072 then return {0} end
-- 所有形状、代次、限额和序列化校验完成后才作唯一写入；不声称Lua错误能回滚先前命令。
redis.call('SET', KEYS[1], changed)
if action == 'release' then return {1,'released',now,now} end
return {1,row.phase,row.expires_ms,now}
